f3/certstore.go

```diff
@@ -0,0 +1,193 @@
+}
+
+var ErrCertAlreadyExists = errors.New("certificate already exists in the CertStore")
+var ErrCertNotFound = errors.New("certificate not found")
+
+// CertStore is responsible for storing and relaying information about new finality certificates
+type CertStore struct {
+	dsLk     sync.Mutex
+	ds       datastore.Datastore
+	busCerts broadcast.Channel[*Cert]
+}
+
+// NewCertStore creates a certstore
+// The passed Datastore has to be thread safe.
+func NewCertStore(ctx context.Context, ds datastore.Datastore) (*CertStore, error) {
+	cs := &CertStore{
+		ds: namespace.Wrap(ds, datastore.NewKey("/certstore")),
+	}
+	latestCert, err := cs.loadLatest(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("loading latest cert: %w", err)
+	}
+	if latestCert != nil {
+		cs.busCerts.Publish(latestCert)
+	}
+
+	return cs, nil
+}
+
+var certStoreLatestKey = datastore.NewKey("/latestCert")
+
+func (cs *CertStore) loadLatest(ctx context.Context) (*Cert, error) {
+	cb, err := cs.ds.Get(ctx, certStoreLatestKey)
+	if errors.Is(err, datastore.ErrNotFound) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, xerrors.Errorf("getting latest cert: %w", err)
+	}
+	var c Cert
+	err = c.UnmarshalBinary(cb)
+	if err != nil {
+		return nil, xerrors.Errorf("unmarshalling latest cert: %w", err)
+	}
+	return &c, err
+}
+
+// Latest returns the newest available certificate
+func (cs *CertStore) Latest() *Cert {
+	return cs.busCerts.Last()
+}
+
+func (cs *CertStore) Get(ctx context.Context, instance uint64) (*Cert, error) {
+	b, err := cs.ds.Get(ctx, cs.keyForInstance(instance))
+
+	if errors.Is(err, datastore.ErrNotFound) {
+		return nil, xerrors.Errorf("cert at %d: %w", instance, ErrCertNotFound)
+	}
+	if err != nil {
+		return nil, xerrors.Errorf("accessing cert in datastore: %w", err)
+	}
+
+	var c Cert
+	if err := c.UnmarshalBinary(b); err != nil {
+		return nil, xerrors.Errorf("unmarshalling cert: %w", err)
+	}
+	return &c, err
+}
+
+// GetRange returns a range of certs from start to end inclusive by instance numbers in the
+// increasing order. Only this order of traversal is supported.
+// If it encounters missing cert, it returns a wrapped ErrCertNotFound and the available certs
+func (cs *CertStore) GetRange(ctx context.Context, start uint64, end uint64) ([]Cert, error) {
+	if start > end {
+		return nil, xerrors.Errorf("start is larger then end: %d > %d", start, end)
+	}
+	bCerts := make([][]byte, 0, end-start+1)
+	i := start
+
+	for ; i <= end; i++ {
+		b, err := cs.ds.Get(ctx, cs.keyForInstance(i))
+		if errors.Is(err, datastore.ErrNotFound) {
+			break
+		}
+		if err != nil {
+			return nil, xerrors.Errorf("accessing cert at %d for range request: %w", i, err)
+		}
+
+		bCerts = append(bCerts, b)
+	}
+
+	certs := make([]Cert, len(bCerts))
+	for j, bCert := range bCerts {
+		err := certs[j].UnmarshalBinary(bCert)
+		if err != nil {
+			return nil, xerrors.Errorf("unmarshalling a cert at i=%d, instance %d: %w", i, start+uint64(i), err)
```

Comment:
```suggestion
			return nil, xerrors.Errorf("unmarshalling a cert at i=%d, instance %d: %w", j, start+uint64(j), err)
```

Reply:
this can get even simpler as now we don't hold the lock getting, so I can get and immediately unmarshal

Reply:
but I I will keep it the way it is